lib/mongo/topology.ex

```diff
@@ -0,0 +1,216 @@
+defmodule Mongo.Topology do
+  use GenServer
+  alias Mongo.Events.{ServerDescriptionChangedEvent, ServerOpeningEvent, ServerClosedEvent,
+                      TopologyDescriptionChangedEvent, TopologyOpeningEvent, TopologyClosedEvent}
+  alias Mongo.TopologyDescription
+  alias Mongo.ServerDescription
+  alias Mongo.Monitor
+
+  @type initial_type :: :unknown | :single | :replica_set_no_primary | :sharded
+
+  @doc ~S"""
+  Starts a new topology connection, which handles pooling and server selection
+  for replica sets.
+
+  ## Options
+
+    * `:database` - **REQUIRED:** database for authentication and default
+    * `:connect_timeout_ms` - maximum timeout for connect
+    * `:seeds` - a seed list of hosts (without the "mongodb://" part) within the
+      cluster, defaults to `["localhost:27017"]`
+    * `:type` - a hint of the topology type, defaults to `:unknown`, see
+      `t:initial_type/0` for valid values
+    * `:set_name` - the expected replica set name, defaults to `nil`
+    * `:heartbeat_frequency_ms` - the interval between server checks, defaults
+      to 10 seconds
+
+  ## Error Reasons
+
+    * `:single_topology_multiple_hosts` - a topology of type :single was set but
+      multiple hosts were given
+    * `:set_name_bad_topology` - a `:set_name` was given but the topology was set
+      to something other than `:replica_set_no_primary` or `:single`
+  """
+  @spec start_link(Keyword.t, Keyword.t) ::
+          {:ok, pid} |
+          {:error, reason :: atom}
+  def start_link(opts, gen_server_opts \\ []) do
+    GenServer.start_link(__MODULE__, opts, gen_server_opts)
+  end
+
+  def connection_for_address(pid, address) do
+    GenServer.call(pid, {:connection, address})
+  end
+
+  def topology(pid) do
+    GenServer.call(pid, :topology)
+  end
+
+  def stop(pid) do
+    GenServer.stop(pid)
+  end
+
+  ## GenServer Callbacks
+
+  # see https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#configuration
+  @doc false
+  def init(opts) do
+    seeds = Keyword.get(opts, :seeds, [
+      Keyword.get(opts, :hostname, "localhost") <> ":" <> Keyword.get(opts, :port, "27017")
+    ])
```

Comment:
This assumes the `:port` key is passed in as a string. This should probably be changed to allow a numerical port.